File: server_integration_tests/java/java_server_reactive/src/main/java/dev/graphlink/javaserver/impl/ArticleServiceImpl.java
package dev.graphlink.javaserver.impl;

import dev.graphlink.javaserver.generated.enums.ArticleType;
import dev.graphlink.javaserver.generated.inputs.CreateArticleInput;
import dev.graphlink.javaserver.generated.inputs.UpdateArticleInput;
import dev.graphlink.javaserver.generated.interfaces.GlArticleProjection;
import dev.graphlink.javaserver.generated.services.ArticleService;
import dev.graphlink.javaserver.generated.types.Article;
import graphql.GraphQLContext;

import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.ArrayList;
import java.util.NoSuchElementException;

@Service
public class ArticleServiceImpl implements ArticleService {

    @Override
    public Mono<Article> getArticle(String id) {
        return Mono.justOrEmpty(Data.articles.stream().filter(a -> a.getId().equals(id)).findFirst().orElse(null));
    }

    @Override
    public Flux<Article> listArticles() {
        return Flux.fromIterable(new ArrayList<>(Data.articles));
    }

    @Override
    public Mono<Article> createArticle(CreateArticleInput input) {
        System.out.println("Creating article");
        var article = Article.builder()
            .id(Data.nextId())
            .title(input.getTitle())
            .authorId(input.getAuthorId())
            .published(false)
            .build();
        Data.articles.add(article);
        Data.articleCreatedSink.tryEmitNext(article);
        return Mono.just(article);
    }

    @Override
    public Mono<Article> updateArticle(UpdateArticleInput input) {
        for (int i = 0; i < Data.articles.size(); i++) {
            var a = Data.articles.get(i);
            if (a.getId().equals(input.getId())) {
                var updated = Article.builder()
                    .id(a.getId())
                    .title(input.getTitle() != null ? input.getTitle() : a.getTitle())
                    .type(a.getType())
                    .authorId(a.getAuthorId())
                    .webSite(a.getWebSite())
                    .published(a.getPublished())
                    .build();
                Data.articles.set(i, updated);
                Data.articleUpdatedSink.tryEmitNext(updated);
                return Mono.just(updated);
            }
        }
        return Mono.error(new NoSuchElementException("Article not found: " + input.getId()));
    }

    @Override
    public Flux<Article> articleCreated() {
        return Data.articleCreatedSink.asFlux();
    }

    @Override
    public Flux<Article> articleUpdated(String id) {
        return Data.articleUpdatedSink.asFlux().filter(a -> a.getId().equals(id));
    }

    @Override
    public Mono<GlArticleProjection> getProjectedArticle(GraphQLContext graphQLContext) {
        GlArticleProjection projection = Data.articles.stream().findFirst().orElse(null);
        return Mono.justOrEmpty(projection);
    }

    @Override
    public Flux<String> articleDeleted() {
        return Flux.just("deleted-1", "deleted-2");
    }

    @Override
    public Flux<ArticleType> getArticleTypes() {
        return Flux.just(ArticleType.NEWS, ArticleType.BLOG, ArticleType.REVIEW);
    }

    @Override
    public Mono<Article> getArticleWithCount() {
        return Mono.justOrEmpty(Data.articles.stream().findFirst().orElse(null));
    }

    @Override
    public Mono<GlArticleProjection> getArticleInfo(GraphQLContext graphQLContext) {
        return Mono.justOrEmpty(Data.articles.stream().findFirst().orElse(null));
    }
}

File: server_integration_tests/kotlin/kotlin_server_suspend/src/main/kotlin/dev/graphlink/kotlinserversuspend/impl/ArticleServiceImpl.kt
package dev.graphlink.kotlinserversuspend.impl

import dev.graphlink.kotlinserversuspend.generated.enums.ArticleType
import dev.graphlink.kotlinserversuspend.generated.inputs.CreateArticleInput
import dev.graphlink.kotlinserversuspend.generated.inputs.UpdateArticleInput
import dev.graphlink.kotlinserversuspend.generated.services.ArticleService
import dev.graphlink.kotlinserversuspend.generated.types.Article
import graphql.GraphQLContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.flowOf
import org.springframework.stereotype.Service

@Service
class ArticleServiceImpl : ArticleService {
    private val articleCreatedFlow = MutableSharedFlow<Article>(extraBufferCapacity = 16)
    private val articleUpdatedFlow = MutableSharedFlow<Article>(extraBufferCapacity = 16)

    override suspend fun getArticle(id: String): Article = Data.articles.first { it.id == id }

    override suspend fun listArticles(): List<Article> = Data.articles.toList()

    override suspend fun createArticle(input: CreateArticleInput): Article {
        val article = Article(
            id = Data.nextId(),
            title = input.title,
            authorId = input.authorId,
            published = false,
        )
        Data.articles.add(article)
        articleCreatedFlow.tryEmit(article)
        return article
    }

    override suspend fun updateArticle(input: UpdateArticleInput): Article {
        val article = Data.articles.first { it.id == input.id }
        val updated = input.title?.let { article.copy(title = it) } ?: article
        Data.articles[Data.articles.indexOf(article)] = updated
        articleUpdatedFlow.tryEmit(updated)
        return updated
    }

    override fun articleCreated(): Flow<Article> = articleCreatedFlow.asSharedFlow()

    override fun articleUpdated(id: String): Flow<Article> =
        articleUpdatedFlow.asSharedFlow().filter { it.id == id }

    override fun articleDeleted(): Flow<String> = flowOf("deleted-1", "deleted-2")

    // Article implements Article, so a concrete article is a valid
    // projection — the client selects whichever subset it wants.
    override suspend fun getProjectedArticle(): Article =
        Data.articles.first()

    override suspend fun getArticleInfo(): Article? =
        Data.articles.firstOrNull()

    override suspend fun getArticleWithCount(): Article? = Data.articles.firstOrNull()

    override suspend fun getArticleTypes(): List<ArticleType> =
        listOf(ArticleType.NEWS, ArticleType.BLOG, ArticleType.REVIEW)
}
